opentreemap/works_management/models.py

```diff
@@ -21,10 +22,28 @@ class Team(models.Model):
 class WorkOrder(Auditable, models.Model):
     instance = models.ForeignKey(Instance)
     name = models.CharField(max_length=255, null=False, blank=False)
+    reference_num = models.IntegerField(default=0)
     created_at = models.DateTimeField(auto_now_add=True)
     created_by = models.ForeignKey(User)
     updated_at = models.DateTimeField(auto_now=True)
 
+    class Meta:
+        unique_together = ('instance', 'reference_num')
+
+    def get_next_reference_num(self):
+        """
+        Return next sequential reference number for instance.
+        """
+        agg_result = Instance.objects.filter(id=self.instance.id) \
+            .annotate(max_value=Max('workorder__reference_num'))[0]
+        max_value = agg_result.max_value or 0
+        return max_value + 1
```

Comment:
> I'm curious to see what happens when multiple workers try to insert tasks at the same time.

This is simple, it is the same thing that happens _outside_ of a transaction: you'll get an `IntegrityError` from the `unique_together` constraint you added :smile: .

Also, `select_for_update` won't catch a race condition when creating the first task for a map, but it can in all other cases. From the Django docs:
> All matched entries will be locked until the end of the transaction block, meaning that other transactions will be prevented from changing or acquiring locks on them.

So if `save_with_user` is wrapped in a `@transaction.atomic` decorator and a call to `Task.objects.filter(instance=self.instance).select_for_update()` is called before the query to get the max value, any concurrent workers should block if necessary while attempting to acquire a lock for the all the pre-existing `Task` rows.